### PR TITLE
Weighted AUROC to omit empty classes

### LINKED ISSUE #348: [metrics] AUROC Metric can't handle 0 observations of a class with multiclass classifier
I'm attempting to calculate AUROC for a multiclass problem where some classes are very rare, occasionally never seen, and I'm getting the following error: `raise ValueError("No positive samples in targets, true positive value should be meaningless")`

In the case of 0 observations, I feel the `average='weighted'` should work, since the contribution to the final AUROC should be 0 regardless. One can think of other scenarios where there are a very high number of classes, some of which will happen to not be seen in some dataset.

_Originally posted by @BeyondTheProof in https://github.com/PyTorchLightning/pytorch-lightning/issues/2210#issuecomment-872440776_

### PR DESCRIPTION
**Pull request by BeyondTheProof**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #348.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/functional/classification/auroc.py [hunk 1]
```diff
@@ -89,6 +90,21 @@ def _auroc_compute(
     else:
         if mode != DataType.BINARY and num_classes is None:
             raise ValueError('Detected input to `multiclass` but you did not provide `num_classes` argument')
+
+        # If one or more classes has 0 observations, we should exclude them, as its weight will be 0 anyway
+        elif average == AverageMethod.WEIGHTED and len(torch.unique(target)) != num_classes:
```
Comment (Borda) [inline torchmetrics/functional/classification/auroc.py:95]:
```suggestion
        if average == AverageMethod.WEIGHTED and len(torch.unique(target)) != num_classes:
```
the above ends with `raise`

**Review by Borda — APPROVED** [review]
lgtm

File: torchmetrics/functional/classification/auroc.py (hunk 1 above)
Comment (BeyondTheProof) [reply torchmetrics/functional/classification/auroc.py]: Good point, will fix!

File: torchmetrics/functional/classification/auroc.py [hunk 2]
```diff
@@ -89,6 +90,21 @@ def _auroc_compute(
...
             raise ValueError('Detected input to `multiclass` but you did not provide `num_classes` argument')
+
+        # If one or more classes has 0 observations, we should exclude them, as its weight will be 0 anyway
+        if average == AverageMethod.WEIGHTED and len(torch.unique(target)) != num_classes:
+            target_bool_mat = torch.zeros((len(target), num_classes), dtype=bool)
+            target_bool_mat[torch.arange(len(target)), target.long()] = 1
+            class_observed = target_bool_mat.sum(axis=0) > 0
+            for c in range(num_classes):
+                if not class_observed[c]:
+                    warnings.warn(f'Class {c} had 0 observations, omitted from AUROC calculation', UserWarning)
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/auroc.py:101]: could we add a test that checks the warning gets raised for some invalid input?

File: torchmetrics/functional/classification/auroc.py [hunk 3]
```diff
@@ -89,6 +90,21 @@ def _auroc_compute(
...
+            preds = preds[:, class_observed]
+            target = target_bool_mat[:, class_observed]
+            target = torch.where(target)[1]
+            num_classes = class_observed.sum()
+            if num_classes <= 1:
+                raise ValueError(f'Found {num_classes} non-empty classes in `multiclass` AUROC calculation')
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/auroc.py:107]: could we add a test that checks the error gets raised for some invalid input?

**Review by SkafteNicki — COMMENTED** [review]
Looking good, remember to also update the changelog as this influences the end user!

File: torchmetrics/functional/classification/auroc.py (hunk 2 above)
Comment (BeyondTheProof) [reply torchmetrics/functional/classification/auroc.py]: Added, thank you for the suggestion!

File: torchmetrics/functional/classification/auroc.py (hunk 3 above)
Comment (BeyondTheProof) [reply torchmetrics/functional/classification/auroc.py]: Done and done!

**Comment by BeyondTheProof** [comment]
@SkafteNicki I've addressed all your suggestions, and my tests are passing locally. However, code coverage is failing, but I don't know how to address this, and whether it is important or not in this scenario. I would appreciate this workflow approved to make sure everything is A-OK! Thanks again.

**Comment by SkafteNicki** [comment]
@BeyondTheProof do not worry about the code coverage. Only test to focus on are the once marked as "required"

**Review by SkafteNicki — APPROVED** [review]
LGTM :]
